crates/txpool/src/lib.rs

```diff
@@ -1,10 +1,116 @@
-// Copyright 2021 Gnosis Ltd.
-// SPDX-License-Identifier: Apache-2.0
+// Copyright 2015-2018 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Generic Transaction Pool
+//!
+//! An extensible and performant implementation of Ethereum Transaction Pool.
+//! The pool stores ordered, verified transactions according to some pluggable
+//! `Scoring` implementation.
+//! The pool also allows you to construct a set of `pending` transactions according
+//! to some notion of `Readiness` (pluggable).
+//!
+//! The pool is generic over transactions and should make no assumptions about them.
+//! The only thing we can rely on is the `Scoring` that defines:
+//!  - the ordering of transactions from a single sender
+//!  - the priority of the transaction compared to other transactions from different senders
+//!
+//! NOTE: the transactions from a single sender are not ordered by priority,
+//! but still when constructing pending set we always need to maintain the ordering
+//! (i.e. `txs[1]` always needs to be included after `txs[0]` even if it has higher priority)
+//!
+//! ### Design Details
+//!
+//! Performance assumptions:
+//! - Possibility to handle tens of thousands of transactions
+//! - Fast insertions and replacements `O(per-sender + log(senders))`
+//! - Reasonably fast removal of stalled transactions `O(per-sender)`
+//! - Reasonably fast construction of pending set `O(txs * (log(senders) + log(per-sender))`
+//!
+//! The removal performance could be improved by trading some memory. Currently `SmallVec` is used
+//! to store senders transactions, instead we could use `VecDeque` and efficiently `pop_front`
+//! the best transactions.
+//!
+//! The pending set construction and insertion complexity could be reduced by introducing
+//! a notion of `nonce` - an absolute, numeric ordering of transactions.
+//! We don't do that because of possible implications of EIP208 where nonce might not be
+//! explicitly available.
+//!
+//! 1. The pool groups transactions from particular sender together
+//!    and stores them ordered by `Scoring` within that group
```

Comment:
Like my previous comment. I would advise to double check the dependencies that txs could make about their own txs with lower nonces